### ISSUE TITLE
Updating 1.19.0 worlds with modded dimensions to 1.19.1 breaks the world

### ISSUE THREAD

**Issue by TelepathicGrunt** [MinecraftForge/MinecraftForge#8915]
The twist is that the dimension must be made with a modded chunk generator or biome source to trigger the game's inability to parse the level.dat file. Attached below is the world in 1.19.0 with Bumblezone mod on, the datapack error when trying to access the world in 1.19.1 without Bumblezone mod on, and the missing nether/end when trying to access the world in 1.19.1 with the Bumblezone mod on. Notice how in the logs, it complains about an unsupported key being Bumblezone's chunk generator. I did not change the name of the modded chunk generator at all. My 1.19.1 update is really just a fix for the music disc items as their constructors changed. I did not change anything regarding worldgen.

1.19.0 world entered with Bumblezone on.
World file: [TestBefore.zip](https://github.com/MinecraftForge/MinecraftForge/files/9215519/TestBefore.zip)
![image](https://user-images.githubusercontent.com/40846040/181669337-ef2bf6fd-edb2-4a56-ae70-4a6c8003b20a.png)

1.19.0 world entered in 1.19.1 without Bumblezone on.
Log file: https://paste.ee/p/1D8pR - [latest.log](https://github.com/MinecraftForge/MinecraftForge/files/9219138/latest.log)
World file: [TestAfter.zip](https://github.com/MinecraftForge/MinecraftForge/files/9215539/TestAfter.zip)
![image](https://user-images.githubusercontent.com/40846040/181669634-e8c7496c-f45c-4deb-a63c-12a7f9b87c36.png)

1.19.0 world entered in 1.19.1 with Bumblezone on.
Log file: https://paste.ee/p/Khql0 - [latest.log](https://github.com/MinecraftForge/MinecraftForge/files/9219131/latest.log)
World file: [TestAfterWithMod.zip](https://github.com/MinecraftForge/MinecraftForge/files/9215573/TestAfterWithMod.zip)
My unreleased 1.19.1 version of my mod: [the_bumblezone_forge-6.1.5+1.19.1.zip](https://github.com/MinecraftForge/MinecraftForge/files/9215568/the_bumblezone_forge-6.1.5%2B1.19.1.zip)
![image](https://user-images.githubusercontent.com/40846040/181670066-ea12d8eb-d689-4fe7-be6e-962e8223c12a.png)

I cannot release my mod while this issue persist as it would cause too many people to freak out about missing vanilla dimensions or broken worlds. Hence this issue report in hopes that this can be patched by Forge to help both my mod and other modded dimension mods

**Comment by TelepathicGrunt** [comment]
Linking past post of a similar form of this issue: https://github.com/MinecraftForge/MinecraftForge/pull/7613

**Comment by TelepathicGrunt** [comment]
The datapack error screen when loading the world without the mod on part is actually a dup of this issue report. My mistake. https://github.com/MinecraftForge/MinecraftForge/issues/8800

So the only new thing now is 1.19 world upgraded to 1.19.1 but keeps the dimension mod on will nuke nether and end dimensions. Most likely because of data fixer upper

**Comment by LexManos** [comment]
Unable to reproduce this using datapack based dimensions. I suspect it is an issue with your custom generators and not being properly registered in 1.19.1.
Unfortunately, without being able to reproduce and not able to do any debugging using your mod there isn't much I can do.

**Comment by TelepathicGrunt** [comment]
The chunk generator and biome source are indeed registered for my mod as seen here:
https://github.com/TelepathicGrunt/Bumblezone/blob/8a76cf4e279047126dd898ea2b4a7f37a3b9bb8b/src/main/java/com/telepathicgrunt/the_bumblezone/modinit/BzDimension.java#L23

https://github.com/TelepathicGrunt/Bumblezone/blob/8a76cf4e279047126dd898ea2b4a7f37a3b9bb8b/src/main/java/com/telepathicgrunt/the_bumblezone/Bumblezone.java#L115

Nothing in Mc has changed how the dimension is registered between 1.19.0 and 1.19.1. I have even asked Twilight Forest devs to try their dimension mod in 1.19.1 using a 1.19.0 world made with their mod and they were able to replicate the issue.

**Comment by TelepathicGrunt** [comment]
Furthermore, if my chunk generator was unregistered in 1.19.1, then I would get this datapack verification failed screen instead and logs about unbounded values (this is in dev with me commenting out my chunk generator deferred registry). So the fact I can access and load in the world means the chunk generator is indeed registered
<img src="https://user-images.githubusercontent.com/40846040/182494757-d67684bd-e83a-4bdb-b618-1b9aef692d9d.png" data-canonical-src="https://user-images.githubusercontent.com/40846040/182494757-d67684bd-e83a-4bdb-b618-1b9aef692d9d.png" width="500" height="300" />

Here's a full video of the issue from start to finish showing how 1.19.0 world with custom modded dimension being loaded into 1.19.1 for first time with same mods will nuke nether and end dimensions. There's nothing more I can do to showcase the issue more than this. This is all I have left.

https://streamable.com/hy9chb

**Comment by Commoble** [comment]
I have reproduced this myself in this small sandbox mod with two classes and a dimension json. To clarify, upgrading the world from 1.19 to 1.19.1 does not prevent the custom dimension from loading in 1.19.1, it prevents the nether and end from existing.
<https://github.com/Commoble/sandbox/tree/8915>

Debug log:
<https://gist.github.com/Commoble/800314983b3d9ad3b1e11ad7bf0af4dc>

Reproduction steps:
* clone repo
* set mc_version to 1.19 and forge_version to 41.1.0 in gradle.properties
* regen IDE runs and refresh gradle in IDE
* run minecraft 1.19 in dev environment
* create and join world
* `/execute in ` to verify existence of overworld, nether, end, and test dimensions
* exit world
* close minecraft
* set mc and forge versions to 1.19.1 and 42.0.1 in gradle.properties
* regen IDE runs and refresh gradle in IDE
* run minecraft 1.19.1 in dev environment
* attempt to join world created in 1.19
* (`[Render thread/ERROR] [mojang/DataFixerUpper]: Unsupported key: sandbox:test` is reported in the log if and only if this is the first time the 1.19 world was joined in 1.19.1, just before the experimental worldgen warning screen displays)
* `/execute in ` only shows overworld and test dimensions (nether and end are missing). Nether and end dimensions do not have entries in level.dat.

Disturbingly, removing the test dimension json and deleting the test dimension entry in level.dat does *not* reconstitute the nether and end dimensions, they seem to be permanently gone for that world.

**Comment by marchermans** [comment]
Okay, so bare with me, this took a very deep dive into DFU to understand why this happens. So let me clarify some points:

_Let's start with some terminology:_
- **Schema**: A collection of Types that the game can understand at a given point in its life (so for a given version)
- **Type**: The definition and layout of a piece of data. This can be any form of data, JSON, NBT, etc. Important here is that a Type defines what it is (List, String, int, a complex object with fields, etc), which properties the data has (the keys in JSON or NBT, etc), and what kind of type is stored in that property (this last bit is important!). For types that represent lists, the Type also defines what its children look like and how to identify them (This is also important later on!).
-  **Fix**: A way to transfer one or more Types from Schema A to Schema B (generally known as the Input and Output schemas)
-  **Rule**: A computationally optimized version of a Fix.

_Now what are the consequences of these definitions:_
1. When Minecraft updates a piece of Data (regardless of Json or NBT again) from one version to the next, it first determines what Type the input, and what the Output is, then it reads with the input and writes with the output. Applying the fixes in the process (note this is a watered-down version, it is possible that it can even do it with the same type under certain conditions (SAME_NAMESPACE schemas are an example!))
2. It can never read anything it is not made aware of (since a property of a type needs to define what type it contains, and a type always needs to define all properties and their layout (aka be complete)).
3. Types that are looked up via their key property (more on that later) need to know all values that this key can take on. Else the type would not be complete and deserialization would fail!

_How does this apply to the current situation:_
- The last schema version in which the WORLD_GEN_SETTINGS type (which represents the data structure of the world gen settings field in saves and its children) was V2832 (All newer schemas always inherit from the last version that modified it, saving code 😄). This schema defines the layout of the settings as follows:
```java
      p_185234_.registerType(false, References.WORLD_GEN_SETTINGS, () -> {
         return DSL.fields("dimensions", DSL.compoundList(DSL.constType(namespacedString()), DSL.fields("generator", DSL.taggedChoiceLazy("type", DSL.string(), ImmutableMap.of("minecraft:debug", DSL::remainder, "minecraft:flat", () -> {
            return DSL.optionalFields("settings", DSL.optionalFields("biome", References.BIOME.in(p_185234_), "layers", DSL.list(DSL.optionalFields("block", References.BLOCK_NAME.in(p_185234_)))));
         }, "minecraft:noise", () -> {
            return DSL.optionalFields("biome_source", DSL.taggedChoiceLazy("type", DSL.string(), ImmutableMap.of("minecraft:fixed", () -> {
               return DSL.fields("biome", References.BIOME.in(p_185234_));
            }, "minecraft:multi_noise", () -> {
               return DSL.or(DSL.fields("preset", namespacedString().template()), DSL.list(DSL.fields("biome", References.BIOME.in(p_185234_))));
            }, "minecraft:checkerboard", () -> {
               return DSL.fields("biomes", DSL.list(References.BIOME.in(p_185234_)));
            }, "minecraft:the_end", DSL::remainder)), "settings", DSL.or(DSL.constType(DSL.string()), DSL.optionalFields("default_block", References.BLOCK_NAME.in(p_185234_), "default_fluid", References.BLOCK_NAME.in(p_185234_))));
         })))));
      });
``` 
- What you need to notice from this is that the generators field in an object in the dimensions list is identified via the generators type field which is a tagged choice (meaning that the value of the key property (here type) is used to figure out the layout of the generator object) and that the values are loaded from the map it is given. Note as well that this map is an ImmutableMap with all chunk generator type names as keys.
- This means that when DFU tries to read in the NBT of a world with ANY custom chunk generator DFU will prevent this from being parsed throw an exception since it has no knowledge of that generator type in its type layout and has no way to parse that.

_Why is this a problem_:
- Well when an exception here is thrown the surrounding list notes down the exception in its result processing state: 
```java
            final MutableObject<DataResult<Unit>> result = new MutableObject<>(DataResult.success(Unit.INSTANCE, Lifecycle.experimental()));

            map.accept((key, value) -> {
                final DataResult<K> k = keyCodec.parse(ops, key);
                final DataResult<V> v = elementCodec.parse(ops, value);

                final DataResult<Pair<K, V>> readEntry = k.apply2stable(Pair::new, v);

                readEntry.error().ifPresent(e -> failed.put(key, value));

                result.setValue(result.getValue().apply2stable((u, e) -> {
                    read.add(e);
                    return u;
                }, readEntry));
            });
```
- As you can see from the above snippet, once an errored result has been put in the `result` field, the `apply2stable` callback will not run anymore. And no new values will be added to the `read` list which keeps track of the resulting entries which are parsed. Additionally, all errored values are added to the `failed` map.
- When the list parser finishes this is then returned as a partial success result (regardless of if there are any, it is always a partial with the ones that were successfully parsed until the first error occurs in the successful entry and all errored ones in the errored section of the partial, the successfully parsed ones between the first error and the end are discarded because they are neither added to the `read` list of `failed` map.).
- The resulting partial result is updated and then attempted to be written, and here it turns out the errored values are added back in with their RAW NBT from the failed parsing and then handed to the NBT Codec for WorldGenSettings.

_How can this be fixed_:
TLDR: That is currently unknown. 

- DFU has all of these classes marked as final and so we can't extend them. A fix for a similar issue with maps instead of lists has been stale for nearly 2 years so we don't have any hope there. We could run a custom DFU version for forge but the consequences of that are completely unclear.
- We could adapt the Schema to teach it how to handle this particular case, but again this is not the only place where this issue crops up, and it will likely not be the only schema that needs fixing. We would rather have a system where modders could modify the schema and its types, but that is currently still under research (again everything in DFU is final so hard to modify).

I hope to have informed you as of know why this is happening and that we are in the process of finding a solution for this.

Greets,

Orion
